### PR TITLE
[RFC] Browserify

### PR DESCRIPTION
**Pull request by ericclemmons**
### Setup
1. `npm install -g browserify`
2. `browserify src/client/app/js/app.js -o dist/js/app.js`
### TODO
- [ ] Discuss, otherwise the rest shouldn't be done.
- [ ] `grunt browserify` task (the existing NPM module sucks).
- [ ] `app.templates` doesn't quite fit in with `require`, since it is compiled.

### CODE REVIEW

**Comment by ericclemmons** [comment]
@kwhitley We talked about if this were possible or not & how clean/ugly it is.  Anyway, in ~30 minutes, the demo still works and I've renamed & restructured files to match CommonJS conventions.

The code is pretty similar:

``` js
app.module('app', [
  'app.controllers.name'              // OLD
  require('./controllers/guide').name // NEW
]);
```

I'm not saying this is good or bad, but simply _interesting_.  Here is my initial list:

**PRO** - There are no assumptions as of the name of the module anymore.  You only have to reference the path.  The module object will be passed around by reference and instantiated only once.

**PRO** - Dependency tree only has to be as large as `require`d (pun intended!).  With `concat`, _everything_, whether the app needs it or not, is combined.  Each folder's `index.js`, for example, may not be necessary.  Unused directives & controllers is another example of optional bloat.

**PRO** - Consistency with server-side coding.

**PRO** - Angular modules can seamlessly use [Components](http://component.io/), since they use `require` as well.

**PRO** - Isolation within (CommonJS) modules.  With `module.exports`, we can write locally-scoped utility classes, functions, etc. _outside_ of the controller/directive/whatever.  I'm not a huge fan of having large directive `link` or `controller` functions, so this would make it possible to export the directive itself, but reference the controller locally to keep it clean. (Hoisting FTW).

**CON** - [Browserify](http://browserify.org/).  Adds bloat, but is minimal (196 bytes!) with the newest version.  

**CON** - Knowledge of CommonJS.  Front-end code typically doesn't use this pattern, so it's kind of _odd_ to see it there.  Not sure how bad this is, since each of these files use `angular`, and are magically combined at the end anyway, so there will be some "magic" no matter what we use...

**CON** - `require('./controllers/guide').name` doesn't look as clean or explicit as `app.controller.guide`, but maybe that's because `require` isn't syntax-highlighted in ST2 for me...

**Comment by kwhitley** [comment]
Curious to take a look at this... Clearly much to discuss!

File: src/client/app/js/controllers/guide.js [hunk 1]
```diff
@@ -0,0 +1,63 @@
+module.exports = guide = angular.module('app.controllers.guide', []);
+
+guide.controller('app.controllers.guide', [
+
+  '$scope',
+
+  function($scope) {
+    var sections = [
+      { id:       "requirements" }
+    , { id:       "installation" }
```
Comment (kwhitley) [inline src/client/app/js/controllers/guide.js:10]: no, no, no, no, NO.

File: src/client/app/js/controllers/guide.js [hunk 2]
```diff
@@ -0,0 +1,63 @@
+module.exports = guide = angular.module('app.controllers.guide', []);
```
Comment (kwhitley) [inline src/client/app/js/controllers/guide.js:1]: What do you think about simplifying this to something like...

``` js
module.exports = angular.module('app.controllers.guide', []);

exports.controller('app.controllers.guide', [
...
```

In theory that should work...

**Comment by kwhitley** [comment]
Hmmm biggest problem with this is maintaining the path relativity.  Cross referencing modules from different parts of the app will have to be very specific with the require paths - and moving folders without consequence will be out of the question.

**Comment by ericclemmons** [comment]
Decided against this because it's a deviation because of...
- Cross-module references
- Overriding modules by namespace
- etc.

**Pull request closed without merging**